data-generation/generate-instructions.py

```diff
@@ -0,0 +1,204 @@
+            self.side2_letter == other.side2_letter    and \
+            self.side1_color == other.side1_color      and \
+            self.side2_color == other.side2_color      or  \
+                                                           \
+            self.side1_letter == other.side2_letter    and \
+            self.side2_letter == other.side1_letter    and \
+            self.side1_color == other.side2_color      and \
+            self.side2_color == other.side1_color
+
+    def __ne__(self, other):
+        return not self == other
+
+    def shift_to(self, position):
+        return Block(self.side1_letter,
+                     self.side1_color,
+                     self.side2_letter,
+                     self.side2_color,
+                     position)
+
+    def flip(self):
+        return Block(self.side2_letter,
+                     self.side2_color,
+                     self.side1_letter,
+                     self.side1_color,
+                     self.position)
+
+    def __str__(self):
+        return '{side1: (%s, %s), side2: (%s, %s), pos: %s}' % \
+            (self.side1_letter, self.side1_color,
+             self.side2_letter, self.side2_color,
+             self.position)
+
+
+class Instruction:
+    def __init__(self, phrase, point):
+        self.phrase = phrase
+        self.point = point
+
+    def __str__(self):
+        return self.phrase
+
+
+class Configuration:
+    def __init__(self, blocks, final_blocks=[]):
+        self.current_blocks = blocks
+        self.final_blocks = final_blocks
+
+    def is_complete(self):
+        return self.current_blocks == []
+
+    def get_instruction(self, moved_block):
+        point = moved_block.position
+        phrase = 'move the %s %s block here: %s' \
+                 % (moved_block.side1_color,
+                    moved_block.side1_letter,
+                    point)
+        return Instruction(phrase, point)
+
+    def get_action(self, goal):
+        if self.is_complete():
+            raise NoActionException('Board is already complete')
+
+        block_to_move = rand_element(self.current_blocks)
+        moved_block = goal.final_blocks[goal.final_blocks.index(block_to_move)]
+        new_current_blocks = self.current_blocks[:]
+        new_current_blocks.remove(moved_block)
+        new_configuration = Configuration(
+            new_current_blocks,
+            self.final_blocks + [moved_block]
+        )
+        instruction = self.get_instruction(moved_block)
+        return Action(self, new_configuration, instruction)
+
+    def scatter(self):
+        return Configuration(list(map(randomize_block,
+                                      self.current_blocks + self.final_blocks)))
+
+    def __str__(self):
+        # :(
+        return str(list(map(str, self.current_blocks + self.final_blocks)))
+
+    def mark_complete(self):
+        return FinalConfiguration(self.current_blocks + self.final_blocks)
+
+
+class FinalConfiguration(Configuration):
```

Comment:
What's the benefit of having this different class? Can't we pass in `[]` for `current_blocks` and the full list for `final_blocks` in the `Configuration` constructor to achieve just this? You could even make both parameters optional if you really wanted to go all-out.

Reply:
I don't think there really is one. I believe I initially made it for argument reordering, and left it because it seemed reasonable to make the final configuration special in some way. I could take it out.